browser/ipfs/ipfs_tab_helper.cc

```diff
@@ -368,11 +468,49 @@ void IPFSTabHelper::DidFinishNavigation(content::NavigationHandle* handle) {
       handle->IsSameDocument()) {
     return;
   }
+  auto is_ipfs_companion_enabled(
+      pref_service_->GetBoolean(kIPFSCompanionEnabled));
+  if (!is_ipfs_companion_enabled &&
+      ((handle->IsErrorPage() && handle->GetNetErrorCode() != net::OK) ||
+       (IsClientOrServerHttpError(handle)))) {
+    auto* orig_url_nav_data = static_cast<IpfsFallbackOrigUrlData*>(
+        handle->GetUserData(kIpfsFallbackOriginalUrlKey));
+    if (orig_url_nav_data && !orig_url_nav_data->original_url.is_empty()) {
+      ShowBraveIPFSFallbackInfoBar(orig_url_nav_data->original_url);
+      return;
+    }
+  }
+
   if (handle->GetResponseHeaders() &&
       handle->GetResponseHeaders()->HasHeader(kIfpsPathHeader)) {
     MaybeSetupIpfsProtocolHandlers(handle->GetURL());
   }
-  MaybeCheckDNSLinkRecord(handle->GetResponseHeaders());
+
+  auto* block_redirect_nav_data = static_cast<IpfsFallbackBlockRedirectData*>(
+      handle->GetUserData(kIpfsFallbackBlockRedirectForNavEntryIdKey));
+  const bool auto_redirect_blocked =
+      block_redirect_nav_data && block_redirect_nav_data->enable_redirect_block;
+
+  MaybeCheckDNSLinkRecord(handle->GetResponseHeaders(), auto_redirect_blocked);
+}
+
+void IPFSTabHelper::ShowBraveIPFSFallbackInfoBar(
+    const GURL& initial_navigation_url) {
+  if (show_fallback_infobar_callback_for_testing_) {
+    show_fallback_infobar_callback_for_testing_.Run(initial_navigation_url);
+    return;
+  }
+#if !BUILDFLAG(IS_ANDROID)
+  auto* content_infobar_manager =
+      infobars::ContentInfoBarManager::FromWebContents(web_contents());
+  if (content_infobar_manager) {
```

Comment:
Interesting - on Android, it doesn't look like we support using a local gateway so I guess we don't need any UI for it :smile:

Reply:
we support remote gateways on Android